myia/ir.py

```diff
@@ -1,7 +1,142 @@
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        output: The `Node` whose value will be returned by this function.
+
+    """
+
+    def __init__(self, output: 'Node') -> None:
+        """Construct a graph."""
+        self.output = output
+        self.debug: Dict = {}
 
 
 class Node:
-    """A node in the graph-based ANF IR."""
+    """A node in the graph-based ANF IR.
+
+    There are three types of nodes: Values, which are the result of function
+    applications; parameters; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a value, the first node input is the function to
+            apply, followed by the arguments.
+        value: The value of this node, if it is a constant.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self) -> None:
+        """Construct a node."""
+        self.inputs: List[Node] = []
+        self.value: LiteralType = None
+        self.graph: Graph = None
+        self.debug: Dict = {}
+
+
+class Value(Node):
+    """A value.
+
+    A value is the result of function application.
+
+    """
+
+    def __init__(self, inputs: List[Node], graph: 'Graph') -> None:
+        """Construct a value."""
+        if len(inputs) < 1:
+            raise ValueError
+        super().__init__()
+        self.inputs = inputs
+        self.graph = graph
+
+
+class Parameter(Node):
+    """A parameter to a function.
+
+    Parameters are leaf nodes, since they are not the result of a function
+    application, and they have no value. They are entirely defined by the graph
+    they belong to.
+
+    """
+
+    def __init__(self, graph: Graph) -> None:
+        """Construct the parameter."""
+        super().__init__()
+        self.graph = graph
+
+
+class Constant(Node):
+    """A constant node.
+
+    A constant is a node which is not the result of a function application. In
+    the graph it is a leaf node. It has no inputs, and instead is defined
+    entirely by its value. Unlike parameters and values, constants do not
+    belong to any particular function graph.
+
+    """
+
+    def __init__(self, value: LiteralType) -> None:
+        """Construct a literal."""
+        super().__init__()
+        self.value = value
+
+
+class Literal(Constant):
+    """A literal.
+
+    A literal is a string or a numeric value.
+
+    """
+
+    def __init__(self, value: Union[str, float, int]) -> None:
+        """Construct a literal."""
+        super().__init__(value)
+
+
+class PrimitiveOperation(Enum):
+    """Built-in primitive operations."""
+
+    ADD = auto()
+    SUB = auto()
+    MULT = auto()
+    DIV = auto()
+
+
+class Primitive(Constant):
+    """A primitive.
```

Comment:
I feel that class is probably superfluous, because I don't think you usually want to do anything general with it: you want to match specific primitives, which works just as well with `Constant`.

Reply:
So there's a few reasons I introduced a class hierarchy, which maybe we can use to figure out which subclasses are needed and which aren't:

* A node should not be allowed to be in an invalid state i.e. have both `inputs` and `value` defined, so those two should probably be separate classes.
* Quick testing of frequent cases `isinstance(n, Primitive)` instead of `isinstance(n, Constant) and isinstance(n.value, Primitive)`.
* Code readability (i.e. `Primitive(ADD)` is more intuitive than `Constant(ADD)`.

The first one I think is important, the second two are readability concerns more than anything, so if you don't think we need those cases very often, I can fold everything into `Apply` and `Constant` nodes.